src/main/java/uk/gov/pay/directdebit/webhook/gocardless/services/handlers/SendEmailsForGoCardlessEventsHandler.java

```diff
@@ -0,0 +1,91 @@
+import uk.gov.pay.directdebit.payments.model.Payment;
+import uk.gov.pay.directdebit.payments.services.PaymentQueryService;
+
+import javax.inject.Inject;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Predicate;
+
+import static uk.gov.pay.directdebit.events.model.GoCardlessResourceType.MANDATES;
+import static uk.gov.pay.directdebit.events.model.GoCardlessResourceType.PAYMENTS;
+import static uk.gov.pay.directdebit.events.model.GoCardlessResourceType.PAYOUTS;
+
+public class SendEmailsForGoCardlessEventsHandler {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SendEmailsForGoCardlessEventsHandler.class);
+    
+    private static Predicate<GoCardlessEvent> byValidEventsForMandateResourceTypes =
+            event -> MANDATES.equals(event.getResourceType());
+    
+    private static Predicate<GoCardlessEvent> byValidEventsForPaymentsAndPayoutsResourceTypes =
+            event -> Set.of(PAYMENTS, PAYOUTS).contains(event.getResourceType()) &&
+                    Set.of("created", "submitted", "confirmed", "failed", "paid_out", "paid")
+                            .contains(event.getAction().toLowerCase());
+
+    private final MandateQueryService mandateQueryService;
+    private final UserNotificationService userNotificationService;
+    private final PaymentQueryService paymentQueryService;
+
+    @Inject
+    public SendEmailsForGoCardlessEventsHandler(MandateQueryService mandateQueryService,
+                                                UserNotificationService userNotificationService,
+                                                PaymentQueryService paymentQueryService) {
+        this.mandateQueryService = mandateQueryService;
+        this.userNotificationService = userNotificationService;
+        this.paymentQueryService = paymentQueryService;
+    }
+
+    public void handle(List<GoCardlessEvent> events) {
+        events.stream().filter(byValidEventsForMandateResourceTypes).forEach(e -> handleMandate(e));
+        events.stream().filter(byValidEventsForPaymentsAndPayoutsResourceTypes).forEach(e -> handlePayment(e));
+    }
+
+    private void handlePayment(GoCardlessEvent event) {
+        logEvent(event);
+        GoCardlessPaymentId goCardlessPaymentId = event.getLinksPayment()
+                .orElseThrow(() -> new GoCardlessEventHasNoPaymentIdException(event.getGoCardlessEventId()));
+
+        Payment payment = paymentQueryService.findByGoCardlessPaymentIdAndOrganisationId(goCardlessPaymentId, event.getLinksOrganisation())
+                .orElseThrow(() -> new PaymentNotFoundException(goCardlessPaymentId, event.getLinksOrganisation()));
+
+        if (event.getAction().equalsIgnoreCase("failed")) {
+            userNotificationService.sendPaymentFailedEmailFor(payment);
+        }
+    }
+
+    private void handleMandate(GoCardlessEvent event) {
+        logEvent(event);
+        switch (event.getAction().toLowerCase()) {
```

Comment:
Again, no need to convert to lower-case.

Reply:
personally i don't mind a bit of defensive programming here but i've changed it anyway